### ISSUE TITLE
macOS: client and/or server is hung

### ISSUE THREAD

**Issue by gorgatron1** [syd711/vpin-studio#445]
**Describe the bug**

Client was hung (thread dump attached), now server appear to be hung-ish -- on launch it is showing the spinner forever while trying to show all tables.

**To Reproduce**

I was playing "Big Game (Stern)" and quit to popper, then went to play "Star Gazer".  I noticed that the high score popup didn't show -- I thought maybe since I had started Star Gazer before it showed it held off (per discussion in #436).  After I quit Star Gazer I waited in Popper but I never saw the score card show up.  I went back to the client to look at the high scores event log and that is when the client hung (I grabbed a thread dump, attached below).

I killed off the client and reran it.  It came up to the point where it is showing the All Tables view, but it can't load the view.  It was able to get the log zip file.

[threads.txt](https://github.com/user-attachments/files/16903208/threads.txt)
[vpin-studio-logs.zip](https://github.com/user-attachments/files/16903209/vpin-studio-logs.zip)

**Expected behavior**

No hang (or whatever it is doing)

**Desktop (please complete the following information):**
 - OS: macOS 15

**Comment by gorgatron1** [comment]
Note: popper is still running at the point where I grabbed the logs but no VPX process.  The restart server button in the settings did work (the client was able to display the list of tables again).

**Comment by gorgatron1** [comment]
I ran Big Game and exited it again and the server/client hung in exactly the same way.  Curiously the nvram does isn't even supported by pinemhi (biggame.nv)

**Comment by syd711** [comment]
"Unfortunately" the issue has nothing todo with the Studio, but was a crash of the Java Virtual Machine (JVM). I'm not sure how stable the JVM we are using is on macOS. If the issue happens more often, we could try to update it.
Of course the server should not be affected by this, but I see no way to reproduce this scenario.
I leave the ticket open if additional crashes happen and will also check how we could update client JVMs.

**Comment by gorgatron1** [comment]
Do you think the client crashed because of the thread dump?  I actually triggered that with a `^\` (SIGQUIT).  It might look like a crash, but it was just me.

I think it was something on the server side because it persisted between launches of the client, but looking at the server logs I didn't see anything out of the ordinary.  That said, it looks like I _can_ reproduce it on my cabinet -- is there a way to turn on more detailed logs?  Maybe I can try getting a thread dump from that.

**Comment by syd711** [comment]
Ah, I see. Then I misread the logs.
Was the event log filled for Star Gazer? It is indeed odd that nothing is logged after this table exit.
The thread dump was a good approach, but you catched the client, not the server process. Otherwise this may have indeed revealed something. So if you try that again, would be much apprechiated. It wasn't a pleasant suprise to get this report details in the first place :)

**Comment by gorgatron1** [comment]
No, after I exited the Big Game table none of the high scores were scanned.  I will try this tonight and get a thread dump from the server side.

**Comment by gorgatron1** [comment]
Dang, now it works fine -- I can't reproduce it any more.  Maybe close this as we don't have sufficient info.  If I encounter something like this again I will get a thread dump from the server side.  

It looks like I need to get the Zulu JDK installed so I have jstack on that machine.

**Comment by gorgatron1** [comment]
OK, I got a hang:

- was playing World Poker Tour (Stern) (and maybe got a high score)
- exited back to popper
- exited popper -- I was planning to add a new table
- in the client I clicked to install the new table (Hang Glider) and I am hung on this screen

<img width="1365" alt="image" src="https://github.com/user-attachments/assets/b2a809e7-fce7-4a8a-ae5b-534c8ad01494">

here is the thread dump from the server:

[threads.txt](https://github.com/user-attachments/files/16914535/threads.txt)

and the log:

[vpin-studio-server.log](https://github.com/user-attachments/files/16914540/vpin-studio-server.log)

**Comment by gorgatron1** [comment]
hopefully this is the same issue that I originally saw -- the good news is it looks like a simple deadlock:

```
Found one Java-level deadlock:
=============================
"/api/v1/games/scores/384":
  waiting to lock monitor 0x00000269761e9a80 (object 0x0000000604d34138, a de.mephisto.vpin.server.games.GameService),
  which is held by "/api/v1/games/scores/561"
"/api/v1/games/scores/561":
  waiting to lock monitor 0x0000026972371f80 (object 0x0000000606221e20, a de.mephisto.vpin.server.highscores.HighscoreService),
  which is held by "Game Exit Thread [X-Men Magneto LE (Stern 2012) VPW 1.0]"
"Game Exit Thread [X-Men Magneto LE (Stern 2012) VPW 1.0]":
  waiting to lock monitor 0x00000269761e9a80 (object 0x0000000604d34138, a de.mephisto.vpin.server.games.GameService),
  which is held by "/api/v1/games/scores/561"
```

Ah, interesting -- I had played X-Men (one of the hung threads) before World Poker Tour.

From the backtraces it looks like it might just be a case of lock order inversion with HighscoreService -> (TournamentChangeListener) -> GameService and GameService -> HighScoreService.  I don't know how easy that is to reproduce, it might be a timing thing, but the cause doesn't look mysterious at least.

**Comment by syd711** [comment]
Thank you for the detailed analysis, couldn't wish for more.
I tried reproducing it via postman, but got no dead locks so far.

The main issue seems to be that the TournamentChangelistener itself again triggers a highscore change. I also identified some other flaws there, so hopefully the issue is fixed with the next patch.

**Comment by syd711** [comment]
Fixed with 3.5.1

Re-open this issue in case the fix did not work.
